### DISCUSSION TITLE
关于某些软件包提供开启 AVX/AVX2 指令集版本的提议

### DISCUSSION THREAD

**Discussion by DuckSoft** [archlinuxcn/repo#2201] [Ideas]
发现 archlinuxcn 里面已经打包了很多模拟器，小到 NES/SNES 模拟器，大到 PSP/PS/PS2/PS3 模拟器等。
对于这类模拟器软件，性能对用户的体验至关重要。

目前，archlinuxcn 打包的所有此类模拟器软件均未开启 AVX/AVX2 指令集。以 Dolphin Emulator 为例，从 archlinuxcn 直接下载的版本（未开启 AVX2）与自行开启 AVX2 指令集编译的版本，实测在同一个游戏中的模拟帧率差距可达 40%。

参考 Arch Linux 官方的打包，有两组包比较典型：一个是 tensorflow/tensorflow-opt，另一个是 python-pytorch/python-pytorch-opt——有无 -opt 后缀的区别在于是否开启了 AVX2 指令集，这样的命名规则可以参考。

Tracking:

- shadowsocks-rust-git (shadowsocks-rust-opt-git)

**Comment by DDoSolitary** [comment]
我觉得好（

**Comment by cuihaoleo** [comment]
坐等 Arch 分出来 x86_64_v3 architecture 就行了：
https://gitlab.archlinux.org/archlinux/rfcs/-/merge_requests/2/diffs

**Comment by CoelacanthusHex** [comment]
Arch devtools 的 x86_64-v3 支持已合并，并且 devtools-archlinuxcn 业已更新到这个版本。
https://gitlab.archlinux.org/archlinux/devtools/-/merge_requests/86
唯一的问题是，在 Arch 的架构里 x86_64-v3 被视作单独的架构，故在官方发布 x86_64-v3 的包之前我们无法应用。

**Reply by SilverRainZ** [reply]
如果开启了 AVX 指令集的支持，是不是意味着 x86_64-v2 及以下的 CPU 都不能跑这个 binary 了？（还是会 fallback 呢？

**Reply by CoelacanthusHex** [reply]
是的，除非软件做了 hwcaps（比如 openssl）
所以目前 Arch 仓库里的方案是使用两个不同的包，将来的方案干脆是视为不同架构（使用不同的仓库）

**Reply by yan12125** [reply]
> Arch devtools 的 x86_64-v3 支持已合并，并且 devtools-archlinuxcn 业已更新到这个版本。

嗯 那個 MR 是在最新版發布之後才合併的，想試的得到 devtools-git (

>  唯一的问题是，在 Arch 的架构里 x86_64-v3 被视作单独的架构，故在官方发布 x86_64-v3 的包之前我们无法应用。

From https://gitlab.archlinux.org/archlinux/rfcs/-/blob/master/rfcs/0002-march.rst,

> It would be preferable if pacman on x86_64_v3 could still install packages from x86_64, particularly for non-Arch repositories that may not want to build for both architectures. This would also allow a transition into x86_64_v3 when firstly [core] gets rebuilt, followed by other repos one at a time. Your friendly pacman developers may be willing to add the ability to specify multiple architectures in pacman.conf.

最後一句提到的 add the ability to specify multiple architectures in pacman.conf 在 pacman 6 已經實現 https://gitlab.archlinux.org/pacman/pacman/-/commit/3179db108a83104d9de6d1d607f55f8118e92160 。或許 archlinuxcn 不必等官方倉庫支援 x86_64_v3。

**Reply by CoelacanthusHex** [reply]
那看起来很好，我们也许可以着手开始支持，目前来看主要有两个问题：
- 我们需要 cherry pick v3 支持
- 我们需要让 lilac 每次更新跑两次构建

**Reply by lilydjwg** [reply]
单独为 v3 跑一只 lilac 吧，就像现在的 alarm 支持那样。然后有需要的包上，any 包或者没啥影响的包都不放过去了。

**Reply by yan12125** [reply]
> 我们需要 cherry pick v3 支持

正說著，devtools又出新版啦 ( https://github.com/archlinuxcn/repo/pull/2818 )。extra-x86_64_v3-build 有點問題，不知道是哪邊出了錯。
```
$ extra-x86_64_v3-build -c
==> Creating chroot for [extra] (x86_64_v3)...
  -> Deleting chroot copy 'root'...
==> Creating install root at /var/lib/archbuild/extra-x86_64_v3/root
==> Installing packages to /var/lib/archbuild/extra-x86_64_v3/root
:: Synchronizing package databases...
 core downloading...
 extra downloading...
 community downloading...
error: failed retrieving file 'core.db' from archlinux.ccns.ncku.edu.tw : The requested URL returned error: 404
error: failed retrieving file 'extra.db' from archlinux.ccns.ncku.edu.tw : The requested URL returned error: 404
error: failed retrieving file 'community.db' from archlinux.ccns.ncku.edu.tw : The requested URL returned error: 404
warning: too many errors from archlinux.ccns.ncku.edu.tw, skipping for the remainder of this transaction
error: failed retrieving file 'core.db' from mirror.archlinux.tw : The requested URL returned error: 404
error: failed retrieving file 'extra.db' from mirror.archlinux.tw : The requested URL returned error: 404
error: failed retrieving file 'community.db' from mirror.archlinux.tw : The requested URL returned error: 404
warning: too many errors from mirror.archlinux.tw, skipping for the remainder of this transaction
error: failed retrieving file 'core.db' from free.nchc.org.tw : The requested URL returned error: 404
error: failed retrieving file 'extra.db' from free.nchc.org.tw : The requested URL returned error: 404
error: failed retrieving file 'community.db' from free.nchc.org.tw : The requested URL returned error: 404
warning: too many errors from free.nchc.org.tw, skipping for the remainder of this transaction
error: failed to synchronize all databases (failed to retrieve some files)
==> ERROR: Failed to install packages to new root
==> ERROR: Failed to install all packages
==> ERROR: Aborting...
```

> 我们需要让 lilac 每次更新跑两次构建

或許仿造 alarmcn 的作法，開一個新的目錄 (ex: archlinuxcn-v3)，有需要的人自行把包複製過去，跑另外一隻 lilac 建構那裡面的包會比較簡單？

**Reply by CoelacanthusHex** [reply]
> > 我们需要 cherry pick v3 支持
> 
> 正說著，devtools又出新版啦 ( #2818 )。extra-x86_64_v3-build 有點問題，不知道是哪邊出了錯。
> 
> ```
> $ extra-x86_64_v3-build -c
> ==> Creating chroot for [extra] (x86_64_v3)...
>   -> Deleting chroot copy 'root'...
> ==> Creating install root at /var/lib/archbuild/extra-x86_64_v3/root
> ==> Installing packages to /var/lib/archbuild/extra-x86_64_v3/root
> :: Synchronizing package databases...
>  core downloading...
>  extra downloading...
>  community downloading...
> error: failed retrieving file 'core.db' from archlinux.ccns.ncku.edu.tw : The requested URL returned error: 404
> error: failed retrieving file 'extra.db' from archlinux.ccns.ncku.edu.tw : The requested URL returned error: 404
> error: failed retrieving file 'community.db' from archlinux.ccns.ncku.edu.tw : The requested URL returned error: 404
> warning: too many errors from archlinux.ccns.ncku.edu.tw, skipping for the remainder of this transaction
> error: failed retrieving file 'core.db' from mirror.archlinux.tw : The requested URL returned error: 404
> error: failed retrieving file 'extra.db' from mirror.archlinux.tw : The requested URL returned error: 404
> error: failed retrieving file 'community.db' from mirror.archlinux.tw : The requested URL returned error: 404
> warning: too many errors from mirror.archlinux.tw, skipping for the remainder of this transaction
> error: failed retrieving file 'core.db' from free.nchc.org.tw : The requested URL returned error: 404
> error: failed retrieving file 'extra.db' from free.nchc.org.tw : The requested URL returned error: 404
> error: failed retrieving file 'community.db' from free.nchc.org.tw : The requested URL returned error: 404
> warning: too many errors from free.nchc.org.tw, skipping for the remainder of this transaction
> error: failed to synchronize all databases (failed to retrieve some files)
> ==> ERROR: Failed to install packages to new root
> ==> ERROR: Failed to install all packages
> ==> ERROR: Aborting...
> ```

感觉像是因为没有 v3 的仓库

> > 我们需要让 lilac 每次更新跑两次构建
> 
> 或許仿造 alarmcn 的作法，開一個新的目錄 (ex: archlinuxcn-v3)，有需要的人自行把包複製過去，跑另外一隻 lilac 建構那裡面的包

这个办法听起来不错

**Reply by yan12125** [reply]
> 感觉像是因为没有 v3 的仓库

我再看了一下那個 pacman patch，他只是允許安裝多個架構的包 (ex: both `sudo pacman -U foo-123-1-x86_64_v3.pkg.tar.zst` and `sudo pacman -U bar-321-1-x86_64.pkg.tar.zst` work)。對於 mirror，會直接用 Architecture= 指定的第一個架構代入`$arch`，如果沒有特別的 hack，的確需要等官方支援 x86_64_v3。

有個 workaround：

```Diff
diff --git a/pacman-extra-x86_64_v3.conf b/pacman-extra-x86_64_v3.conf
index afb7a0c..385c091 100644
--- a/pacman-extra-x86_64_v3.conf
+++ b/pacman-extra-x86_64_v3.conf
@@ -19,7 +19,7 @@ HoldPkg     = pacman glibc
 #XferCommand = /usr/bin/curl -L -C - -f -o %o %u
 #XferCommand = /usr/bin/wget --passive-ftp -c -O %o %u
 #CleanMethod = KeepInstalled
-Architecture = x86_64_v3 x86_64
+Architecture = x86_64 x86_64_v3
 
 # Pacman won't upgrade packages listed in IgnorePkg and members of IgnoreGroup
 #IgnorePkg   =
```
x86_64 和 x86_64_v3 應該是 ABI-compatible，所以就用 x86_64 的 dependency 來建 x86_64_v3 的包。如果這聽起來合理，我就把他加到 https://github.com/archlinuxcn/repo/pull/2818 。

**Reply by CoelacanthusHex** [reply]
这听起来合理，在 pacman 同时允许安装这两个架构的包本身就隐含了 ABI-compatible 了，否则这一定会导致问题

**Reply by yan12125** [reply]
Thanks! Updated #2818

**Reply by lilydjwg** [reply]
还有个问题： archrepo2 尚未加入 x86_64_v3 的支持。这个支持加起来简单，但是需要把 any 包链接到 x86_64_v3 中吗？

**Reply by yan12125** [reply]
> 需要把 any 包链接到 x86_64_v3 中吗

應該不需要，加了也沒壞處。

在可以預期的未來，archlinuxcn 只會有一部分的包打 x86_64_v3，如果使用者不是只用這些 v3 的包，就需要同時開啟 x86_64 和 x86_64_v3 兩個倉庫。例如：
```
[archlinuxcn-v3]
Server = https://repo.archlinuxcn.org/x86_64_v3
[archlinuxcn]
Server = https://repo.archlinuxcn.org/x86_64
```
(兩個倉庫應該得用不同名字，這可能需要再討論)

需要的 any 包可從 https://repo.archlinuxcn.org/x86_64 下載。

如果有一天，Arch Linux 官方準備拋棄 x86_64，只保留 x86_64_v3，archlinuxcn 也打算跟進，archlinuxcn-v3 就需要 any 包的連結了。不過那應該也是好幾年之後的事了。

**Reply by lilydjwg** [reply]
要用不同的名字啊。不知道直接 ln -s archlinuxcn.db archlinuxcn-v3.db 行不行？

**Reply by CoelacanthusHex** [reply]
理论上可行，因为 db 没有包含仓库本身的信息
